src/opt/ImproveMemoryOps.cpp

```diff
@@ -653,226 +653,155 @@ static bool lOffsets32BitSafe(llvm::Value **offsetPtr, llvm::Instruction *insert
-               g->target->hasVecPrefetch() ? __pseudo_prefetch_write_varying_2_native : __prefetch_write_varying_2,
-               false, true),
-
-        GSInfo(__pseudo_prefetch_write_varying_3,
-               g->target->hasVecPrefetch() ? __pseudo_prefetch_write_varying_3_native : __prefetch_write_varying_3,
-               g->target->hasVecPrefetch() ? __pseudo_prefetch_write_varying_3_native : __prefetch_write_varying_3,
-               false, true),
-    };
-
-    int numGSFuncs = sizeof(gsFuncs) / sizeof(gsFuncs[0]);
-    for (int i = 0; i < numGSFuncs; ++i)
-        Assert(gsFuncs[i].func != nullptr && gsFuncs[i].baseOffsetsFunc != nullptr &&
-               gsFuncs[i].baseOffsets32Func != nullptr);
-
-    GSInfo *info = nullptr;
-    for (int i = 0; i < numGSFuncs; ++i)
-        if (gsFuncs[i].func != nullptr && callInst->getCalledFunction() == gsFuncs[i].func) {
-            info = &gsFuncs[i];
-            break;
+    GSInfo(const char *pgbo, const char *pgfbo, const char *pgbo32, const char *pgfbo32, GSInfo::Type type)
+        : pgbo(pgbo), pgfbo(pgfbo), pgbo32(pgbo32), pgfbo32(pgfbo32), type(type) {}
+    GSInfo(const char *pgbo, const char *pgfbo, GSInfo::Type type)
+        : pgbo(pgbo), pgfbo(pgfbo), pgbo32(pgbo), pgfbo32(pgfbo), type(type) {}
+    static GSInfo Gather(const char *pgbo, const char *pgfbo, const char *pgbo32, const char *pgfbo32) {
+        return GSInfo(pgbo, pgfbo, pgbo32, pgfbo32, Type::Gather);
+    }
+    static GSInfo Gather(const char *pgbo, const char *pgfbo) { return GSInfo(pgbo, pgfbo, Type::Gather); }
+    static GSInfo Scatter(const char *pgbo, const char *pgfbo, const char *pgbo32, const char *pgfbo32) {
+        return GSInfo(pgbo, pgfbo, pgbo32, pgfbo32, Type::Scatter);
+    }
+    static GSInfo Scatter(const char *pgbo, const char *pgfbo) { return GSInfo(pgbo, pgfbo, Type::Scatter); }
+    static GSInfo Prefetch(const char *pgbo, const char *pgfbo) { return GSInfo(pgbo, pgfbo, Type::Prefetch); }
+    bool isGather() const { return type == GSInfo::Type::Gather; }
+    bool isScatter() const { return type == GSInfo::Type::Scatter; }
+    bool isPrefetch() const { return type == GSInfo::Type::Prefetch; }
+    bool use() const {
+        switch (type) {
+        case GSInfo::Type::Gather:
+            return g->target->useGather();
+        case GSInfo::Type::Scatter:
+            return g->target->useScatter();
+        case GSInfo::Type::Prefetch:
+            return g->target->hasVecPrefetch();
         }
-    if (info == nullptr)
+        return false;
+    }
+    llvm::Function *baseOffsetsFunc() const {
+        const char *name = use() ? pgbo : pgfbo;
+        return m->module->getFunction(name);
+    }
+    llvm::Function *baseOffsets32Func() const {
+        const char *name = use() ? pgbo32 : pgfbo32;
+        return m->module->getFunction(name);
+    }
+    const char *pgbo;
+    const char *pgfbo;
```

Comment:
I would add "64" suffix to these struct members similar to `pgbo32`/`pgfbo32`.
Same for `baseOffsetsFunc`.

Reply:
Fixed